Develop

woobottle opened the pull request:
1. useKeyboard, useModalMessage hook의 경우 쓰임새가 적합한지 궁금합니다. 

dispatch를 전달받아 커스텀 함수를 반환해주는 역할의 훅입니다. 제가 기존에 알고 있는 커스텀 훅의 역할은 컴포넌트의 상태와 관련된 메소드들을 분리하여 관리하는 것으로 알고 있습니다. 허나 지금은 상태가 아닌 dispatch만 넘겨받아 커스텀 메소드들을 리턴해줍니다. 함수 분리의 역할만 수행한다 생각이 되어 질문드리고 싶습니다. 제가 작성한 것과 같이 사용하는 것은 어떠하신지가 궁금합니다.

2. 전체적인 구조에 대해 의견을 여쭤보고 피드백을 받고 싶습니다. 

현재 대략적인 구조는 <App><GameBoard/><Keyboard/></App>  입니다. keyboard의 입력이 app에 정의된 상태를 변화 시키고 gameboard는 리렌더링 됩니다. 애니메이션을 추가 하려고 하기 전에 더 나은 구조가 있을지 의견 여쭤보고 싶습니다.

3. 혹시 저와 다른 방법으로 구현 하셨다면 그에 대한 방법과 현재 방법에서 우려되는 점이나 나아가야 할 점이 있다면 알고 싶습니다. 

아래의 내용은 현재 구현된 방법 입니다.
Game의 상태중 columnIndex, currentInput은 개발의 편의성을 위해 상태로 정의 하였습니다.
만약 gameColumn(단어 하나하나 입력되는 부분)을 input 으로 정의했다면 ref값을 사용하여 비제어 컴포넌트로 작업을 진행할 수 있을것 같다 생각하였지만 focus를 이동 시키는 부분, disabled 설정등의 로직이 복잡할 것으로 예상되어 div와 상태 값들의 조합으로 화면을 출력해줄 수 있도록 하였습니다. 

4. 본인의 저장소 주소를 남겨주시면 꼭 가서 참조할 수 있도록 하겠습니다!!

woobottle committed:
add styled && 본격적 시작

woobottle committed:
add default keyboard ui

woobottle committed:
add useReducer for row, column Index

woobottle committed:
버튼 동작, enter 동작 추가

woobottle committed:
add useCallback to useKeyboard hook

woobottle committed:
add columnIndex and currentInput

woobottle committed:
add button color setting

woobottle committed:
add keyboard background setting

woobottle committed:
add localStorage setting

woobottle committed:
edit localStorage logic

woobottle committed:
keymapper 위치 이동
key 입력 비교를 위한 keyMapper는 렌더링 마다 생성될 필요가 없으므로 함수형 컴포넌트 외부로 이동

woobottle committed:
edit readme

woobottle committed:
try to refactoring useKeyboard

woobottle committed:
seperate useKeyboard to useGame

woobottle committed:
feat: modal edit 시작

woobottle committed:
util로 뺄 수 있는것들은 분리

woobottle committed:
feat: add modal animation but need more code & actions

woobottle committed:
fix: remove, add 동작 버그 처리

woobottle committed:
refactor: GameBoard 컴포넌트 추가

woobottle committed:
fix: edit background color logic

woobottle committed:
edit: result modal logic

woobottle committed:
edit: result modal css

woobottle committed:
edit readme

woobottle committed:
add : add play video

woobottle committed:
edit: reademe contents

woobottle committed:
edit: reademe contents

woobottle committed:
Update README.md

woobottle committed:
edit: reademe contents

woobottle committed:
Merge branch 'develop' of github.com:woobottle/wordle-challenge into develop

developer-1px commented on README.md:
```diff
@@ -18,4 +18,105 @@ yarn
+   => 테스트는 어떤식으로 하게 될까 궁금하네? -> 한번 짜보자
+
+6) keyboard 입력버튼을 ref로 관리, 입력, 제거를 기존에서 분리, currentInput이 상태값으로 존재할 필요가 없다
+   => 실패, 의견을 한번 여쭤보자 
+
+7) 렌더링이 적은게 무조건 좋은걸까? -> 코드가 난해해지면 유지보수가 안된다.
+8) enter를 눌렀을때 rowIndex를 늘리는 것이 자연스러운가, 검사를 하고 나서 올리는게 자연스러운가 (검사를 하고 올리는 것이 맞다)
+  => 엔터의 역할은 단어의 업데이트
+  => 검사의 역할은 입력된 단어의 유효성을 판단, 다음 인덱스로 안내하는 것
+   
+9) 엔터 누르고 단어평가하고 평가된 걸로 검사해야함, 얘네는 분리가 안되는것 같은데 
+    => 값을 사용하는 쪽에서 넣어주고 업데이트만 쳐주자
+    => 둘의 시점이 다르다 엔터를 하고 나서 업데이트 된 애를 검사해야 하는데 이전 것을 계속 검사하고 있었음
+
+10) 모달 출력 로직의 복잡화.... 
+  => 상태하나를 추가하고 해당 상태의 상황에 따라 출력로직 정리
+  => isGameComplete 추가
+ 
+## 구현영상 
+![](May-13-2022%2019-25-24.gif)
+
+
+## 이번 pr에서 받고 싶은 내용들 
+1) useKeyboard, useModalMessage hook의 경우 쓰임새가 적합한지 궁금하니다. 
+  dispatch를 전달받아 커스텀 함수를 반환해주는 역할의 훅입니다.
+  제가 기존에 알고 있는 커스텀 훅의 역할은 컴포넌트의 상태와 관련된 메소드들을 분리하여 관리하는 것으로 알고 있습니다.
+  허나 지금은 상태가 아닌 dispatch만 넘겨받아 커스텀 메소드들을 리턴해줍니다. 
+  함수 분리의 역할만 수행한다 생각이 되어 질문드리고 싶습니다.
+  제가 작성한 것과 같이 사용하는 것은 어떠하신지가 궁금합니다.
+
+2) 전체적인 구조에 대해 의견을 여쭤보고 피드백을 받고 싶습니다.
+  현재 대략적인 구조는 <App><GameBoard/><Keyboard/></App> 입니다.
+  keyboard의 입력이 app에 정의된 상태를 변화 시키고 gameboard는 리렌더링 됩니다.
+  애니메이션을 추가 하려고 하기 전에 더 나은 구조가 있을지 의견 여쭤보고 싶습니다.
+
+3) Game의 상태중 columnIndex, currentInput은 개발의 편의성을 위해 상태로 정의 하였습니다.
+   만약 gameColumn(단어 하나하나 입력되는 부분)을 input 으로 정의했다면 ref값을 사용하여 비제어 컴포넌트로 작업을 진행할 수 있을것 같다 생각하였지만 focus를 이동 시키는 부분, disabled 설정등의 로직이 복잡할 것으로 예상되어 div와 상태 값들의 조합으로 화면을 출력해줄 수 있도록 하였습니다.
+   혹시 저와 다른 방법으로 하셨다면 그에 대한 방법과 현재 방법에서 우려되는 점이나 나아가야 할 점이 있다면 알고 싶습니다.
+
+4) 본인의 저장소 주소를 남겨주시면 꼭 가서 참조할 수 있도록 하겠습니다!!
```
와! 이걸 문서로 남기는 것만으로도 이미 훌륭하네요!!

developer-1px commented on README.md:
```diff
@@ -18,4 +18,105 @@ yarn
+
+
+## 이번 pr에서 받고 싶은 내용들 
+1) useKeyboard, useModalMessage hook의 경우 쓰임새가 적합한지 궁금하니다. 
+  dispatch를 전달받아 커스텀 함수를 반환해주는 역할의 훅입니다.
+  제가 기존에 알고 있는 커스텀 훅의 역할은 컴포넌트의 상태와 관련된 메소드들을 분리하여 관리하는 것으로 알고 있습니다.
+  허나 지금은 상태가 아닌 dispatch만 넘겨받아 커스텀 메소드들을 리턴해줍니다. 
+  함수 분리의 역할만 수행한다 생각이 되어 질문드리고 싶습니다.
+  제가 작성한 것과 같이 사용하는 것은 어떠하신지가 궁금합니다.
+
```
일반적으로 UI 개발을 하게 되면 MVC의 패턴을 따르고 V가 JSX라면 M와 C를 일반적으로 hook으로 분리를 하게 됩니다. 
특히 이렇게 hook으로 분리한다는 의미는 여기의 상태와 상태를 변화시키는 동작들의 범위를 고립시켜서 재사용과 독립성을 확보하기 위한 의미입니다.

useKeyboard과 useModalMessage의 경우 지금 useGame과의 종속성이 있는 상태라서 독립적이라고 보기가 어렵습니다.

modalMessages과 같은 state는 특정 컴포넌트에서만 사용이 되므로 useGame에 있는 modalMessages의 state를 그냥 useModalMessage hook으로 보내서 [state와 action]을 모두 리턴을 받을 수 있도록 작성해서 hook의 종속성을 제거하는 방향으로 가는 것을 추천드립니다.

hooks
- useGame
- useKeyboard
- useModalMessage 
- useTheme

이런식으로 구성을 했다면 useGame와 useKeyboard와는 계층관계는 맞게 성립이 되어야 할거에요~

woobottle replied:
소중한 의견 감사합니다!!

덕분에 고민하고 있던 내용 까지 같이 해결이 되었어요!! 
modalMessages 상태는 옮기고 useModalMessage에 useReducer추가, resetGame은 useGame 내부로 이동시킬 예정입니다!

커스텀 훅의 독립성은 고려하지 못하였던것 같네요. 
파일(useGame)이 커지더라도 내부로 이동시켜 관리하는 방향으로도 수정 진행 해야겠어요 ㅎㅎ

woobottle replied:
> 일반적으로 UI 개발을 하게 되면 MVC의 패턴을 따르고 V가 JSX라면 M와 C를 일반적으로 hook으로 분리를 하게 됩니다. 특히 이렇게 hook으로 분리한다는 의미는 여기의 상태와 상태를 변화시키는 동작들의 범위를 고립시켜서 재사용과 독립성을 확보하기 위한 의미입니다.
> 
> useKeyboard과 useModalMessage의 경우 지금 useGame과의 종속성이 있는 상태라서 독립적이라고 보기가 어렵습니다.
> 
> modalMessages과 같은 state는 특정 컴포넌트에서만 사용이 되므로 useGame에 있는 modalMessages의 state를 그냥 useModalMessage hook으로 보내서 [state와 action]을 모두 리턴을 받을 수 있도록 작성해서 hook의 종속성을 제거하는 방향으로 가는 것을 추천드립니다.
> 
> hooks
> 
> * useGame
> * useKeyboard
> * useModalMessage
> * useTheme
> 
> 이런식으로 구성을 했다면 useGame와 useKeyboard와는 계층관계는 맞게 성립이 되어야 할거에요~

혹시 useGame과 useKeyboard와의 계층관계를 맞게 성립이 되어야 한다는 부분에서 설명을 좀 더 부탁드릴수 있을까요??

제가 이해한건 폴더로 구성해서 useGame/useModalMessage 이런식으로 계층 관계를 표현해야 하는것으로 이해해서 맞게 이해했는지가 궁금합니다

developer-1px replied:
useGame이 커지는게 우려가 된다면 useKeyboard등을 useGame에서 import 하는 방식으로 관리를 하는 것도 괜찮습니다.

developer-1px replied:
useKeyboard는 useGame의 서브 모듈이자 종속관계를 가집니다. 직접 import를 하는 강력한 결합은 아니고 dispatch를 이용한 느슨한 결합으로 되어 있는데 지금 경우에서는 둘을 느슨하게 분리를 해야할 이유는 잘 보이지 않습니다. 

지금 작성하신 파일 구조의 의미는 동일한 키보드 액션을 재사용할 수 있도록 해서 useGame1, useGame2처럼 같은 키보드 액션을 가지고 다른 형태의 게임로직을 연결할 수 있도록 만든 형상인데 지금의 어플리케이션은 useGame안에 useKeboard가 있어도 될 만큼 작으니까요.

hooks 폴더 아래에 같은 계층으로 배치한만큼 useGame와 useKeyboard는 동등하거나 독립적인 관계이어야 하나 지금 그렇지 못하기 때문에 구조를 변경하는게 좋아보인다는 의미로 말씀드렸습니다.

woobottle replied:
아 이해하였습니다.  상세한 설명 감사합니다!!!

파일의 구조, 같은 계층에 대해서도 생각을 해보게 되네요 ㅎㅎ 감사합니다!!!
전달해주신 피드백들 반영한 후에 버전 2 pr로 찾아뵙겠습니다!!

developer-1px commented on README.md:
```diff
@@ -18,4 +18,105 @@ yarn
+  함수 분리의 역할만 수행한다 생각이 되어 질문드리고 싶습니다.
+  제가 작성한 것과 같이 사용하는 것은 어떠하신지가 궁금합니다.
+
+2) 전체적인 구조에 대해 의견을 여쭤보고 피드백을 받고 싶습니다.
+  현재 대략적인 구조는 <App><GameBoard/><Keyboard/></App> 입니다.
+  keyboard의 입력이 app에 정의된 상태를 변화 시키고 gameboard는 리렌더링 됩니다.
+  애니메이션을 추가 하려고 하기 전에 더 나은 구조가 있을지 의견 여쭤보고 싶습니다.
```
하나씩 제거 해봤을때 의미적으로 충분히 잘 분리되는 것 같아서 좋은 것 같아요.

woobottle committed:
edit: feedback
game state에서 modalMessages 삭제,
resetGame 적절한 위치로 이동
useGame과 useKeyboard의 느슨한 결합 해제,
useModalMessage도 종속성 해제

같은 계층에 위치한 파일들은 동등하거나 독립적인 관계를 유지하기 위해
변경

InSeong-So commented on src/App.tsx:
```diff
@@ -1,25 +1,58 @@
-import React from 'react';
-import logo from './logo.svg';
 import './App.css';
```
`App.css` 내부의 CSS가 사용되는 게 아니라면 내부 내용을 합치거나 제거하는 게 좋을 것 같네요 :)
해당 PR의 파일에는 올라와 있지 않지만, `index.css`, `App.test.tsx` 등 사용하지 않는 파일도 제거해주시는 게 좋습니다!

woobottle replied:
불필요한 파일의 제거 확인했습니다!!

InSeong-So commented on src/utils/index.ts:
```diff
@@ -0,0 +1,80 @@
+  if (gameStatus === GAME_STATUS.FAIL) {
+    return answer;
+  }
+  if (rowIndex === 0) {
+    return "Genius";
+  }
+  if (rowIndex === 1) {
+    return "Magnificent";
+  }
+  if (rowIndex === 2) {
+    return "Impressive";
+  }
+  if (rowIndex === 3) {
+    return "Splendid";
+  }
+  if (rowIndex === 4) {
+    return "Great";
+  }
```
lookup table에 대한 내용을 참고해서 적용하면 좋겠네요!

woobottle replied:
혹시 말씀해주신 내용에 대해 좀 더 설명이 가능하실까요???

array로 선언해서 ["Genius", "Magnificent", "Impressive", "Splendid", "Great"] 와 같이 선언해서 접근을 하는 것을 혹시 말씀하시는 것일지 알고싶어요

InSeong-So replied:
```ts
const lookupTable = {
	'0': 'Genius',
	'1': 'Magnificent',
	'2': 'Impressive',
	'3': 'Splendid',
	'4': 'Great'
};

return lookupTable[rowIndex];
```

ㅎㅎ 실제 검색을 통해 얻는 인사이트가 클 것 같아 구체적으로 사례를 말씀드리지 않은 것이지만, 적어둘게요 :)
객체의 키-값 쌍이 1:1 매핑되는 특징을 이용해 다중 if문을 효과적으로 제거하는 방식입니다. 가독성도 높아지고 추가적인 로직 정의를 하기 쉬워지죠. 자바스크립트로 계산기를 만드는 튜토리얼들이 웹에 많은데 이를 통해 전체 앱을 구성하는 Top-down 학습도 좋으나, 로직/함수 같은 세부 기술은 Bottom-up 학습을 추천드립니다!

woobottle replied:
감사합니다!!  학습 방법도 확인했씁니다!

InSeong-So commented on src/App.tsx:
```diff
@@ -1,25 +1,58 @@
-        </a>
-      </header>
-    </div>
+    <ThemeProvider theme={{ currentTheme, setTheme }}>
+      <GlobalStyles />
+      <NavModalPortal>
+        {modalMessages.reverse().map((message) => 
+          <NavFadeModal
+            fadeTime={3000}
+            message={message.message}
+            callback={() => removeMessage(message.id)}
+          />)
+        }
+      </NavModalPortal>
+      <div className="App">
+        <GameBoard 
+          {...state} />
+        <KeyBoard
+          {...state}
+          {...keyBoardActions}
+          addMessage={addMessage}
+        />
+      </div>
+      <ResultModalPortal>
+        {isResultModalOpen && 
+          <ResultModal 
+            {...state}
+            callback={() => keyBoardActions.resetGame()}
+          />
+        }
+      </ResultModalPortal>
+    </ThemeProvider>
```
엔트리 파일로 사용하는 `App.tsx`에 없어도 되는 로직이 너무 많은 것 같은데, react folder structure 등을 찾아 적용해보면 어떨까요?

woobottle replied:
react folder structure 탐색 시작하겠습니다!!

InSeong-So commented on src/assets/css/modal.css:
```diff
@@ -0,0 +1,11 @@
+#game-nav-modal {
+  position: fixed;
+  top: 5rem;
+  left: 33%;
+}
+
+#game-result-modal {
+  position: fixed;
+  top: 10rem;
+  left: 33%;
+}
```
styled-components 를 사용하는데 modal에 대한 스타일을 분리해야 할까요?

woobottle replied:
portal을 적용할때에 
<img width="622" alt="image" src="https://user-images.githubusercontent.com/50283326/168454513-ff35d116-0b16-4d2d-868e-38cfefaea73c.png">
이런식으로 작성성했었어서 스스로 그냥 css를 적용해야지 라고 판단을 해버린것 같아요

방법을 찾지 않앗었던게 창피하군여....ㅜㅜ

InSeong-So replied:
[요기](https://github.com/styled-components/styled-components/issues/2620) 참고해보셔요ㅎㅎ

woobottle replied:
감사합니다!!

InSeong-So commented on src/components/GameBoard.tsx:
```diff
@@ -0,0 +1,39 @@
+}: Props) => {
+  return (
+    <GameBoardWrapper>
+      {words.map((word, wordIndex) => {
+        let contents = word.split('');
+        const contentsValidate = wordsEvaulated[wordIndex]
+        if (wordIndex === rowIndex) {
+          contents = currentInput;
+        }
+        
+        return <GameRow contents={contents} contentsValidate={contentsValidate} />
+      })}
```
엘리먼트를 정의하는 항목에 인라인으로 작성된 로직이 너무 많은데, 분리하는 게 좋아보이네요!

woobottle replied:
네 동의합니다! 
별도의 함수로 빼서 동작할 수 있도록 해야겠습니다!

InSeong-So commented on src/components/KeyBoard.tsx:
```diff
@@ -0,0 +1,161 @@
+    return true;
+  };
+  
+  const clickHandler = (e: React.SyntheticEvent) => {
```
`React.SyntheticEvent` 이벤트는 바닐라 자바스크립트의 Event 타입과 동일한 정도로 범위가 넓어요. 타입 좁히기를 찾아보면 좋겠습니다!

woobottle replied:
감사합니다!! MouseEvent로 스코프 줄일수 있도록 할게요!!

InSeong-So commented on src/components/KeyBoard.tsx:
```diff
@@ -0,0 +1,161 @@
+    const flattenWords = words.join('').split('')
+    flattenWords.forEach((char, index) => {
+      if (char === '') return
+      const value = wordsEvaulated[~~(index / WORD_LENGTH)][index % WORD_LENGTH];
```
어떤 것에 대한 연산식인지, value인지 감이 안잡히는데 캡슐화하는 건 어떨까요?

woobottle replied:
동의합니다! 고려하지 못했던 부분이네요 감사합니다!!

InSeong-So commented on src/components/KeyBoard.tsx:
```diff
@@ -0,0 +1,161 @@
+
+    if (buttonValue === '<') {
+      clickDeleteButton()
+      return
```
return 만 정의된 줄에는 세미콜론이 없는데, 통일 시키는 게 좋아보여요!

woobottle replied:
eslint, prettier 당장 적용하도록 할게요 ㅎㅎㅎ

InSeong-So commented on src/hooks/useModalMessage.ts:
```diff
@@ -0,0 +1,26 @@
+const useModalMessage = () => {
+  const [modalMessages, setModalMessages] = useState<ModalMessage[]>([])
+  
+  const addMessage = (message: ModalMessage) => {
+    setModalMessages(prev => [...prev, message])
+  }
```
hook의 느낌보다는 모듈에 가깝네요. prev가 무엇인지, 어떤 역할이 되는지 잘 이해를 못하는 상황입니다ㅠ

woobottle replied:
맞습니다. 해당 부분도 수정진행할게요!!

InSeong-So commented on src/components/NavFadeModal.tsx:
```diff
@@ -0,0 +1,51 @@
+}
+
+const FadeModalWrapper = styled.div<{ time?: number }>`
+  @keyframes fadeout {
+    10% {
+      opacity: 1;
+    }
+    90% {
+      opacity: 0.5;
+    }
+    100% {
+      opacity: 0;
+    }
+  }
```
index.css나 글로벌한 스타일링에 추가하면 안 되는 부분일까요?

woobottle replied:
아 해당 부분은 NavFadeModal 에서만 쓰일것 같아 우선은 해당 컴포넌트 내부에 위치시킨 이후에 추후에 필요성이 생기면 위치를 다시 고민해보는게 어떨까 했었어요!!

InSeong-So reviewed (commented):
코드 잘 봤습니다!
타입과 인터페이스, 로직들이 한 파일에 공존하다보니 중복되는 타입이 퍼져 있는 느낌이 많아요. 또한 네이밍 부분에서 임시 변수 등에 스스로와 타협하지 않았나 하는 부분도 있습니다.
추가적으로 eslint, prettier를 잘 설정하면 좋을 것 같아요! 어디에는 EOL 처리가 안되어있고, 세미콜론 처리가 안되어있고, 스페이스가 균일하지 않는 등... 일관된 코드를 작성하는데 툴의 도움을 받으면 좋겠습니다.
또한 hooks가 사실상 모듈의 역할을 하는 것 같네요. 어디에 어떻게 작성하는게 좋은가, 로직이 컴포넌트의 라이프사이클에 어떻게 필요한가를 고민해보면 좋을 것 같아요!

woobottle commented:
> 코드 잘 봤습니다! 타입과 인터페이스, 로직들이 한 파일에 공존하다보니 중복되는 타입이 퍼져 있는 느낌이 많아요. 또한 네이밍 부분에서 임시 변수 등에 스스로와 타협하지 않았나 하는 부분도 있습니다. 추가적으로 eslint, prettier를 잘 설정하면 좋을 것 같아요! 어디에는 EOL 처리가 안되어있고, 세미콜론 처리가 안되어있고, 스페이스가 균일하지 않는 등... 일관된 코드를 작성하는데 툴의 도움을 받으면 좋겠습니다. 또한 hooks가 사실상 모듈의 역할을 하는 것 같네요. 어디에 어떻게 작성하는게 좋은가, 로직이 컴포넌트의 라이프사이클에 어떻게 필요한가를 고민해보면 좋을 것 같아요!

우선 정말 코드 하나하나 상세히 봐주셔서 감사합니다!!
스스로 타협한 부분이 있었는데 그 부분은 창피하네요 ㅎㅎ 
소중한 의견 감사합니다!!
피드백은 바로 반영작업 시작할게요!!
